### PR TITLE
move new_zeros to core from THP

### PR DESCRIPTION
**Pull request by v0dro**
Fix for issue https://github.com/pytorch/pytorch/issues/25831

@ezyang can you please have a look?

### CODE REVIEW

**Comment by v0dro** [comment]
I checked the failing test and they fail with an error which is on the lines of `Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:339 not found`.

Can you please explain why this would happen?

File: .gitignore [hunk 1]
```diff
@@ -230,6 +230,11 @@ CTAGS
 tags
 TAGS
 
+# Files generated by GNU Global (gtags)
```
Comment (resistor) [inline .gitignore]: Is this intended?  If it is, it should probably be in a separate PR.

**Review by resistor — DISMISSED** [review]

File: .gitignore (hunk 1 above)
Comment (v0dro) [reply .gitignore]: OK I'll send a separate PR. Will update my commit soon.

**Comment by ezyang** [comment]
There were some outages recently, which is why CI failed. Retriggered.

**Review by ezyang — APPROVED** [review]
Waiting on CI

**Comment by v0dro** [comment]
@ezyang any idea why the CI is failing? It seems to a build system issue.

**Comment by ezyang** [comment]
It's an infra problem. Can you rebase onto master? Looks like you have conflicts so I cannot do this automatically.

**Comment by v0dro** [comment]
@ezyang Its done.

**Comment by v0dro** [comment]
@ezyang  Can we merge this?

**Comment by ezyang** [comment]
sorry, there's too many failures, I need to get a clean build of this PR on a non-broken master. sorry about the delay

**Comment by ezyang** [comment]
Build failure is real. Please resolve it as instructed in the logs:

```
Sep 30 14:31:57 --------------------------------------------------------------------------------
Sep 30 14:31:57 caffe2/CMakeFiles/ATEN_CPU_FILES_GEN_TARGET.dir/build.make:104: recipe for target 'aten/src/ATen/CPUType.cpp' failed
Sep 30 14:31:57 CMakeFiles/Makefile2:7580: recipe for target 'caffe2/CMakeFiles/ATEN_CPU_FILES_GEN_TARGET.dir/all' failed
Sep 30 14:31:57 [ 34%] Building CXX object third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/cpu/gemm/s8x8s32/simple_gemm_s8s8s32.cpp.o
Sep 30 14:31:57 [ 34%] Building CXX object third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/cpu/gemm/s8x8s32/jit_avx512_core_u8_copy_sum_bn_kern.cpp.o
Sep 30 14:31:57 [ 34%] Building CXX object third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/cpu/gemm/s8x8s32/jit_avx512_core_u8_copy_sum_bt_kern.cpp.o
Sep 30 14:31:57 [ 34%] Building CXX object third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/cpu/gemm/s8x8s32/jit_avx512_core_gemm_s8u8s32_kern.cpp.o
Sep 30 14:31:57 Traceback (most recent call last):
Sep 30 14:31:57   File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 472, in <module>
Sep 30 14:31:57     generate_outputs()
Sep 30 14:31:57   File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 464, in generate_outputs
Sep 30 14:31:57     "set environment variable GEN_TO_SOURCE or run \"{}\"".format(mismatch, update_cmd))
Sep 30 14:31:57 RuntimeError: Source files: ['TensorMethods.h'] did not match generated files.  To update the source files, set environment variable GEN_TO_SOURCE or run "cp /var/lib/jenkins/workspace/build/aten/src/ATen/core_tmp/TensorMethods.h /var/lib/jenkins/workspace/cmake/../aten/src/ATen/core"
Sep 30 14:31:57 make[2]: *** [aten/src/ATen/CPUType.cpp] Error 1
```

**Comment by v0dro** [comment]
So I had set `GEN_TO_SOURCE=0` on my machine which was causing the build failure. Setting it to `1` has resolved the previous issue that @ezyang pointed out.

However I'm now getting a new [failure](https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-build/48261/) with the following trace:
```
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 472, in <module>
    generate_outputs()
  File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 464, in generate_outputs
    "set environment variable GEN_TO_SOURCE or run \"{}\"".format(mismatch, update_cmd))
RuntimeError: Source files: ['TensorMethods.h'] did not match generated files.  To update the source files, set environment variable GEN_TO_SOURCE or run "cp /var/lib/jenkins/workspace/build/aten/src/ATen/core_tmp/TensorMethods.h /var/lib/jenkins/workspace/cmake/../aten/src/ATen/core"
gmake[2]: *** [caffe2/CMakeFiles/generate-torch-sources.dir/build.make:201: aten/src/ATen/CPUType.cpp] Error 1
gmake[2]: *** Deleting file 'aten/src/ATen/CPUType.cpp'
gmake[1]: *** [CMakeFiles/Makefile2:6153: caffe2/CMakeFiles/generate-torch-sources.dir/all] Error 2
gmake: *** [Makefile:139: all] Error 2
```

Upon scrolling up in trace I find this particular issue:
```
[ 32%] Built target arraymap
Scanning dependencies of target libprotoc
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/code_generator.cc.o
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/command_line_interface.cc.o
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/cpp/cpp_enum.cc.o
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/cpp/cpp_enum_field.cc.o
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/cpp/cpp_extension.cc.o
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/cpp/cpp_field.cc.o
[ 32%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/cpp/cpp_file.cc.o
--------------------------------------------------------------------------------
x=TensorMethods.h, a=/var/lib/jenkins/workspace/build/aten/src/ATen/core_tmp, b=/var/lib/jenkins/workspace/cmake/../aten/src/ATen/core
 1046: -     at::AutoNonVariableTypeMode _var_guard(true);
--------------------------------------------------------------------------------
```

Can someone please tell me why exactly this happens? Would be great to know the cause of this, since the build passes on my machine and I'd like to careful about the build system for future patches.

**Comment by ezyang** [comment]
It sounds like you got unlucky and someone merge conflicted with you again on master. Rebase on master and then run the same thing. (The ROCm builds are weird: they merge with master. Other builds don't.)

(Yes, I know this is horrible. Fix for the situation in #27086)

**Comment by v0dro** [comment]
@ezyang OK this is more successful passes than ever before :P
